esp/services/ldapenvironment/ldapenvironmentService.cpp

```diff
@@ -0,0 +1,204 @@
+
+        xpath.clear().appendf("%ssharedGroupsBaseDN", prefix.str());
+        cfg->getProp(xpath.str(), sharedGroupsBaseDN);
+        if (sharedGroupsBaseDN.isEmpty())
+            throw MakeStringException(-1, "sharedGroupsBaseDN must be specified in configuration (ex. 'ou=groups,ou=shared,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+        xpath.clear().appendf("%ssharedUsersBaseDN", prefix.str());
+        cfg->getProp(xpath.str(), sharedUsersBaseDN);
+        if (sharedUsersBaseDN.isEmpty())
+            throw MakeStringException(-1, "sharedUsersBaseDN must be specified in configuration (ex. 'ou=users,ou=shared,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+        xpath.clear().appendf("%ssharedResourcesBaseDN", prefix.str());
+        cfg->getProp(xpath.str(), sharedResourcesBaseDN);
+        if (sharedResourcesBaseDN.isEmpty())
+            throw MakeStringException(-1, "sharedResourcesBaseDN must be specified in configuration (ex. 'ou=smc,ou=espservices,ou=shared,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+        xpath.clear().appendf("%ssharedWorkunitsBaseDN", prefix.str());
+        cfg->getProp(xpath.str(), sharedWorkunitsBaseDN);
+        if (sharedWorkunitsBaseDN.isEmpty())
+            throw MakeStringException(-1, "sharedWorkunitsBaseDN must be specified in configuration (ex. 'ou=workunits,ou=shared,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+    }
+    catch (...)
+    {
+        throw MakeStringException(-1, "ldapenvironment: Error querying environment settings");
+    }
+}
+
+bool CldapenvironmentEx::onLDAPQueryDefaults(IEspContext &context, IEspLDAPQueryDefaultsRequest &req, IEspLDAPQueryDefaultsResponse &resp)
+{
+    resp.setHPCCRootName(ldapRootOU.str());
+    resp.setAdminGroupName(adminGroupName.str());
+    resp.setSharedFilesBaseDN(sharedFilesBaseDN.str());
+    resp.setSharedGroupsBaseDN(sharedGroupsBaseDN.str());
+    resp.setSharedUsersBaseDN(sharedUsersBaseDN.str());
+    resp.setSharedResourcesBaseDN(sharedResourcesBaseDN.str());
+    resp.setSharedWorkunitsBaseDN(sharedWorkunitsBaseDN.str());
+    resp.setRetcode(0);
+    return true;
+}
+
+const char * CldapenvironmentEx::formatOUname(StringBuffer &ou, const char * envName, int mode, const char * sharedOU, const char * reqBaseDN, const char * privateOU)
+{
+    switch (mode)
+    {
+        case 0: // Generated Private OU
+            ou.appendf("%s,ou=%s,ou=%s", privateOU, envName, ldapRootOU.str());
+            break;
+        case 1: // Global/Shared OU
+            ou.append(sharedOU);//specified in ldapconfiguration
+            break;
+        case 2: // Custom User-provided OU
+            ou.append(reqBaseDN);
+            break;
+    }
+    return ou.str();
+}
+
+bool CldapenvironmentEx::onLDAPCreateEnvironment(IEspContext &context, IEspLDAPCreateEnvironmentRequest &req, IEspLDAPCreateEnvironmentResponse &resp)
+{
+    try
+    {
+        if (secmgr == nullptr)
+        {
+            secmgr = dynamic_cast<CLdapSecManager*>(context.querySecManager());
```

Comment:
secmgr will need to be protected to make it thread-safe.

Reply:
I'm not sure this is the case.  LDAPEnvironment will have it's own instance of the security manager and is the only process on the pod using that instance.  Please elaborate

Reply:
I just mean that onLDAPCreateEnvironment() could be called multiple times at once, so you could have two threads in this block of code at the same time so secmgr should be protected.

Another option that would only happen once at startup might be to have your binding class  (CldapenvironmentSoapBindingEx) override the addService() method.  The binding object inherits a pointer to the secmgr so you can tie everything together there. 

Something like:
```

class CldapenvironmentEx : public Cldapenvironment  //base class name built from ECM ESPservice name
{
public:
   void setSecMgr( ISecManager*sm){secmgr = dynamic_cast<CLdapSecManager*>(sm);}
}

class CldapenvironmentSoapBindingEx : public CldapenvironmentSoapBinding  
{
public:
    void addService(const char * name, const char * host, unsigned short port, IEspService & service) override
    {
        CldapenvironmentEx *srv = dynamic_cast<CldapenvironmentEx*>(&service);
        srv->setSecMgr(querySecManager());  //call EspHttpBinding::querySecManager();
        CEspBinding::addService(name, host, port, service);
    }
};
```

Reply:
Done, using the binding method you recommended